src/main.rs

```diff
@@ -71,203 +84,227 @@ impl Player {
     }
 }
 
-//////////
-// ROOM //
-//////////
-
-#[derive(Debug, PartialEq)]
-enum Danger {
-    Wumpus,
+/// Things that can be in a room.
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum RoomContents {
+    Empty,
     Bat,
     Pit,
 }
+use RoomContents::*;
 
-#[derive(Default, Debug)]
-struct Room {
-    id: usize,
-    neighbours: [Cell<Option<usize>>; ROOM_NEIGHBOURS],
-    dangers: Vec<Danger>,
-}
-
-impl Room {
-    fn new(id: usize) -> Self {
-        let default_room = Room::default();
-
-        Room {
-            id,
-            ..default_room
-        }
-    }
-
-    fn missing_neighbours(&self) -> usize {
-        self.neighbours.iter().filter(|n| n.get().is_none()).count()
-    }
-
-    fn neighbour_ids(&self) -> Vec<usize> {
-        self.neighbours.iter()
-            .filter(|n| n.get().is_some())
-            .map(|n| n.get().unwrap())
-            .collect()
+impl Default for RoomContents {
+    fn default() -> Self {
+        Empty
     }
 }
 
-//////////
-// MAZE //
-//////////
+/// Room description.
+#[derive(Default, Clone, Copy)]
+struct Room {
+    /// The indices of neighboring rooms.
+    neighbors: [usize; ROOM_NEIGHBORS],
+    /// Possible danger in the room.
+    contents: RoomContents,
+    /// Wumpus in the room.
+    wumpus: bool,
```

Comment:
I don't see `wumpus` as a property of the room. Why would the Wumpus be a property, but pits or bats don't? From a "code elegance" point of view, they're all properties, or none of them is, but making them properties would prevent adding a fourth kind of danger to the game later on.

Also, what if I wanted to implement a bigger maze with two Wumpus running free in it?

The Wumpus is a creature and it roams through the map representing a danger just like bats or pits, i don't see why it would be the property of a room.

Reply:
The bats and pits are fixed-in-place and don't move, and it is guaranteed that at most one of them is in a room. The Wumpus moves, and can be in a room with bats or pits. (That's an explicit part of the original rules.) This is the easiest way to model these facts, I think?

Alternative designs: (1) HashSet of room contents — I had that originally and backed it out because it was a bunch of extra code and brought in `HashSet`. (2) Wumpus represented by a location variable, like the player.

Reply:
If you want to change this, I'd be happy to implement alternative (2): I had a hard time deciding which I liked better between that and what I did.

Reply:
> From a "code elegance" point of view, they're all properties, or none of them is, but making them properties would prevent adding a fourth kind of danger to the game later on.

I think the only restriction the current model has is that a room can have at most one danger in it. (The Wumpus is not that kind of danger anymore.) Adding a new danger would be only a few lines of code in the new codebase: extend the enum and fix up the places where it is used. Allowing multiple dangers in a room would mean specifying all the interactions between them, including potentially complicated rules about which dangers could coexist in a room: I wouldn't do this lightly.